shapes/main.cpp

```diff
@@ -1,96 +1,127 @@
+    if (shape) {
+        std::cout << "\n*** " << x << " ***\n";
+        return shape->getArea() < x;
+    }
     return false;
+};
+
+void printMapCollection(const MapCollection& mapCollection) {
+    for (const auto& [shape, perimeter] : mapCollection) {
+        if (shape) {
+            shape->print();
+        }
+    }
 }
 
-void printCollection(const Collection& collection)
-{
-    for (const auto & it : collection)
+void printCollection(const Collection& collection) {
+    for (const auto& it : collection)
         if (it)
             it->print();
 }
 
-void printAreas(const Collection& collection)
-{
-    for (const auto & it : collection)
+void printAreas(const Collection& collection) {
+    for (const auto& it : collection)
         if (it)
             cout << it->getArea() << std::endl;
 }
 
 void findFirstShapeMatchingPredicate(const Collection& collection,
-                                     bool (*predicate)(shared_ptr<Shape> s),
-                                     std::string info)
-{
+                                     std::function<bool(std::shared_ptr<Shape>)> predicate,
+                                     std::string info) {
     auto iter = std::find_if(collection.begin(), collection.end(), predicate);
-    if(*iter != nullptr)
-    {
+    if (*iter != nullptr) {
         cout << "First shape matching predicate: " << info << endl;
         (*iter)->print();
-    }
-    else
-    {
+    } else {
         cout << "There is no shape matching predicate " << info << endl;
     }
 }
 
 constexpr int fibo(int n) {
-    if (n<=2) {
+    if (n <= 2) {
         return 1;
     } else {
         return fibo(n - 1) + fibo(n - 2);
     }
 }
 
-int main()
-{
-    constexpr int n = fibo(45);
-    Collection shapes {
-        make_shared<Circle>(2.0),
-        make_shared<Circle>(3.0),
-        nullptr,
-        make_shared<Circle>(4.0),
-        make_shared<Rectangle>(10.0, 5.0),
-        make_shared<Square>(3.0),
-        make_shared<Circle>(4.0),
-    };
+template <typename T, typename = typename std::enable_if_t<std::is_base_of<Shape, T>::value>>
+void AddShapeToCollection(Collection& collection, std::shared_ptr<T> shape) {
+    collection.emplace_back(shape);
+}
+
+template <class T, class... Arguments>
+std::shared_ptr<Shape> make_shape(Arguments&&... args) {
+    return make_shared<T>(std::forward<Arguments>(args)...);
+}
+
+void FillShapesCollection(Collection& shapes) {
+    AddShapeToCollection(shapes, make_shape<Circle>(2.0));
+    AddShapeToCollection(shapes, make_shape<Circle>(3.0));
+    //AddShapeToCollection(shapes, nullptr); compile error
+    shapes.emplace_back(nullptr);
+    AddShapeToCollection(shapes, make_shape<Circle>(4.0));
+    AddShapeToCollection(shapes, make_shape<Rectangle>(10.0, 5.0));
+    AddShapeToCollection(shapes, make_shape<Square>(3.0));
+    AddShapeToCollection(shapes, make_shape<Square>(4.0));
+}
+
+void FillMapCollection(Collection& shapes, MapCollection& mapCollection) {
```

Comment:
```suggestion
void FillMapCollection(const Collection& shapes, MapCollection& mapCollection) {
```
The first param should not be changed.